File: backend/database/supabase.py
import os

from supabase import Client, create_client

PLACEHOLDER_VALUES = {
    "your-supabase-url",
    "your-supabase-service-key",
}


class SupabaseConfigurationError(RuntimeError):
    pass


def get_supabase_client() -> Client:
    url = os.getenv("SUPABASE_URL", "").strip()
    service_key = os.getenv("SUPABASE_SERVICE_KEY", "").strip()

    if not url or not service_key or url in PLACEHOLDER_VALUES or service_key in PLACEHOLDER_VALUES:
        raise SupabaseConfigurationError("Missing Supabase configuration")

    return create_client(url, service_key)

File: frontend/lib/server/supabase.ts
import { createClient, type SupabaseClient } from "@supabase/supabase-js";

import { getServerEnv } from "./env";

let client: SupabaseClient | null | undefined;

export function getSupabaseServiceClient(): SupabaseClient | null {
  if (client !== undefined) return client;

  const url = getServerEnv("SUPABASE_URL") ?? getServerEnv("NEXT_PUBLIC_SUPABASE_URL");
  const serviceKey = getServerEnv("SUPABASE_SERVICE_KEY");

  client = url && serviceKey ? createClient(url, serviceKey) : null;
  return client;
}
